Undo_Redo_Example.c

```diff
@@ -0,0 +1,298 @@
+#define EFL_BETA_API_SUPPORT 1
+
+#include <Eina.h>
+#include <Elementary.h>
+#include <Efl_Ui.h>
+
+/******************** Undo/Redo logic ********************/
+
+#define MAX_HISTORY_SIZE 5
+
+static Eina_List *history_index = NULL;
+static Eina_List *history_list = NULL;
+static Eina_Bool history_list_can_merge = EINA_FALSE;
+
+static Efl2_Text_Change_Info *
+_history_next_get(void)
+{
+   int size = eina_list_count(history_list);
+
+   if (0 == size || !eina_list_next(history_index))
+     {
+        return NULL;
+     }
+
+   history_index = eina_list_next(history_index);
+
+   return (Efl2_Text_Change_Info *)eina_list_data_get(history_index);
+}
+
+static Efl2_Text_Change_Info *
+_history_prev_get(void)
+{
+   int size = eina_list_count(history_list);
+
+   if (0 == size || !eina_list_prev(history_index))
+     {
+        return NULL;
+     }
+
+   history_index = eina_list_prev(history_index);
+
+   return (Efl2_Text_Change_Info *)eina_list_data_get(history_index);
+}
+
+static void
+_history_append(Efl2_Text_Change_Info *info)
+{
+   int size = eina_list_count(history_list);
+   Efl2_Text_Change_Info *new_info = NULL;
+
+   if (size == MAX_HISTORY_SIZE)
+     {
+        return;
+     }
+
+   if (!new_info)
+        return;
+
+   Eina_List *ll;
+   history_list = eina_list_split_list(history_list, history_index,
+                                       &ll);
+
+   Efl2_Text_Change_Info *inf = NULL;
+   EINA_LIST_FREE(history_list, inf)
+     {
+        eina_stringshare_del(inf->content);
+        free(inf);
+     }
+
+   new_info = (Efl2_Text_Change_Info *)calloc(1, sizeof(*new_info));
+   if (!new_info)
+        return;
+
+   memcpy(new_info, info, sizeof(*new_info));
+
+   Efl2_Text_Change_Info *head_inf = eina_list_data_get(history_index);
+
+   eina_stringshare_ref(new_info->content);
+
+   if (history_list_can_merge && (new_info->insert == head_inf->insert))
+        new_info->merge = EINA_TRUE;
+
+   history_index = eina_list_append(history_index, new_info);
+
+
+   history_index = eina_list_last(history_index);
+   if (history_list == NULL)
+     {
+        history_list = history_index;
+     }
+
+   history_list_can_merge = EINA_TRUE;
+}
+
+static void
+_user_undo(Eo *ui_text)
+{
+   Efl2_Text_Change_Info *info = _history_prev_get();
+
+   while (info)
+     {
+        _user_undo_redo_do(ui_text, info, EINA_FALSE);
+
+        if (info->merge)
+          {
+             info = _history_prev_get();
+          }
+        else
+          {
+             break;
+          }
+     }
+}
+
+
+static void
+_user_undo_redo_do(Eo *ui_text, Efl2_Text_Change_Info *inf, Eina_Bool undo)
+{
+   printf("%s: %s", (undo) ? "Undo" : "Redo",
+         inf->content);
+
+   Efl_Text_Cursor *mcur = efl2_text_raw_editable_main_cursor_get(ui_text);
+   efl2_text_cursor_position_set(mcur, inf->position);
+
+   if ((inf->insert && undo) || (!inf->insert && !undo))
+     {
+        Efl_Text_Cursor  *end;
+        end = efl2_ui_text_cursor_new(ui_text);
```

Comment:
There is no such function.

Reply:
Also, maybe consider caching this cursor? No reason to allocate it every time.

Reply:
Have it open in front of me... Maybe you're based on an old version?

The correct way for creating new cursors is a bit weird at the moment, look at the tests (recent changes in the git log) to see how it's done.

Reply:
```
cursor_new {
         [[Creates and returns a new cursor for the text.]]
         return: ptr(Efl.Text_Cursor_Cursor); [[Text cursor]]
      }
```
I see this inside https://github.com/Enlightenment/efl/blob/devs/tasn/ifaces/src/lib/elementary/efl2_ui_text.eo

Reply:
I was indeed looking at the wrong one! I was looking at canvas_text. Anyhow, this one is an old leftover as you can see from the fact that it doesn't use the correct return value (Efl.Text_Cursor_Cursor is the old type). I'll push a fix in a bit.